### ISSUE TITLE
How to mix and train the different datasets?

### ISSUE THREAD

**Issue by LyazS** [CalciferZh/minimal-hand#47]
Paper say that: _DetNet is trained on 3 datasets:  theCMU Panoptic Dataset (CMU) ,  the Rendered Hand-pose Dataset (RHD) and the GANerated Hands Dataset(GAN)._

Since the images of three datasets are different from each other, can u please tell me how to preprocess the image?

**Comment by CalciferZh** [comment]
Sorry but I'm afraid I don't quite understand your meaning by "the images are different" :)

**Comment by LyazS** [comment]
> Sorry but I'm afraid I don't quite understand your meaning by "the images are different" :)

I mean the distribution of the images from different dataset are different. 
How do u deal with that?

**Comment by CalciferZh** [comment]
As our goal was not to get the best result on some specific dataset but to obtain a model that generalizes well to real world applications, we would like to see that we are using different datasets with different distributions. Therefore, we did not take care of the different distribution at all.

On the other hand, do you have any statistics that can evaluate the image distribution of each dataset? We would like to see how  they are different, if possible.

**Comment by MengHao666** [comment]
> Paper say that: _DetNet is trained on 3 datasets: theCMU Panoptic Dataset (CMU) , the Rendered Hand-pose Dataset (RHD) and the GANerated Hands Dataset(GAN)._
> 
> Since the images of three datasets are different from each other, can u please tell me how to preprocess the image?

Hi, have you knew how the author train the model with different datasets. It seems impossible to train network jointly with all data, as the lables are different and could not shuffle in batch convention.

**Comment by LyazS** [comment]
> 
> 
> > Paper say that: _DetNet is trained on 3 datasets: theCMU Panoptic Dataset (CMU) , the Rendered Hand-pose Dataset (RHD) and the GANerated Hands Dataset(GAN)._
> > Since the images of three datasets are different from each other, can u please tell me how to preprocess the image?
> 
> Hi, have you knew how the author train the model with different datasets. It seems impossible to train network jointly with all data, as the lables are different and could not shuffle in batch convention.

I don't know. The three datasets all have labels wrt 21 keypoints, then I just mix all the img and lab into one dataloader to train the network. But, seems not work. Maybe the RGB distribution or the data distribution or the number of them are different. :(

**Comment by MengHao666** [comment]
> > > Paper say that: _DetNet is trained on 3 datasets: theCMU Panoptic Dataset (CMU) , the Rendered Hand-pose Dataset (RHD) and the GANerated Hands Dataset(GAN)._
> > > Since the images of three datasets are different from each other, can u please tell me how to preprocess the image?
> > 
> > 
> > Hi, have you knew how the author train the model with different datasets. It seems impossible to train network jointly with all data, as the lables are different and could not shuffle in batch convention.
> 
> I don't know. The three datasets all have labels wrt 21 keypoints, then I just mix all the img and lab into one dataloader to train the network. But, seems not work. Maybe the RGB distribution or the data distribution or the number of them are different. :(

It seems  to work  as it is only the implementation problem. I am trying to re-implement author's paper in Pytorch framwork.

**Comment by LyazS** [comment]
> 
> 
> > > > Paper say that: _DetNet is trained on 3 datasets: theCMU Panoptic Dataset (CMU) , the Rendered Hand-pose Dataset (RHD) and the GANerated Hands Dataset(GAN)._
> > > > Since the images of three datasets are different from each other, can u please tell me how to preprocess the image?
> > > 
> > > 
> > > Hi, have you knew how the author train the model with different datasets. It seems impossible to train network jointly with all data, as the lables are different and could not shuffle in batch convention.
> > 
> > 
> > I don't know. The three datasets all have labels wrt 21 keypoints, then I just mix all the img and lab into one dataloader to train the network. But, seems not work. Maybe the RGB distribution or the data distribution or the number of them are different. :(
> 
> It seems to work as it is only the implementation problem. I am trying to re-implement author's paper in Pytorch framwork.

I only implement DetNet with 2D keypoints part. And DetNet with 3D part works very bad, and I don't know the train strategy. If you have re-implement, please share in github. :)

**Comment by MengHao666** [comment]
> > > > > Paper say that: _DetNet is trained on 3 datasets: theCMU Panoptic Dataset (CMU) , the Rendered Hand-pose Dataset (RHD) and the GANerated Hands Dataset(GAN)._
> > > > > Since the images of three datasets are different from each other, can u please tell me how to preprocess the image?
> > > > 
> > > > 
> > > > Hi, have you knew how the author train the model with different datasets. It seems impossible to train network jointly with all data, as the lables are different and could not shuffle in batch convention.
> > > 
> > > 
> > > I don't know. The three datasets all have labels wrt 21 keypoints, then I just mix all the img and lab into one dataloader to train the network. But, seems not work. Maybe the RGB distribution or the data distribution or the number of them are different. :(
> > 
> > 
> > It seems to work as it is only the implementation problem. I am trying to re-implement author's paper in Pytorch framwork.
> 
> I only implement DetNet with 2D keypoints part. And DetNet with 3D part works very bad, and I don't know the train strategy. If you have re-implement, please share in github. :)

I find its good performance when train whole Detnet  exclusively on RHD and STB dataset. It seems tricky because of the preprocess , network and final global alignment, but it is not severly hard.. Now, I am trying to train jointly with 3 datasets jointly  and test on DO /EO dataset like the paper. I will share it if it works well at last with author's permission.
